pkg/externalLinks/ExternalLinkService.go

```diff
@@ -0,0 +1,311 @@
+	for _, request := range requests {
+		t := &ExternalLink{
+			Name:                         request.Name,
+			Active:                       true,
+			ExternalLinkMonitoringToolId: request.MonitoringToolId,
+			Url:                          request.Url,
+			AuditLog:                     sql.AuditLog{CreatedOn: time.Now(), CreatedBy: userId, UpdatedOn: time.Now(), UpdatedBy: userId},
+		}
+		err := impl.externalLinkRepository.Save(t)
+		if err != nil {
+			impl.logger.Errorw("error in saving link", "data", t, "err", err)
+			err = &util.ApiError{
+				InternalMessage: "external link failed to create in db",
+				UserMessage:     "external link failed to create in db",
+			}
+			return nil, err
+		}
+
+		for _, clusterId := range request.ClusterIds {
+			externalLinksMapping := &ExternalLinkClusterMapping{
+				ExternalLinkId: t.Id,
+				ClusterId:      clusterId,
+				Active:         true,
+				AuditLog:       sql.AuditLog{CreatedOn: time.Now(), CreatedBy: userId, UpdatedOn: time.Now(), UpdatedBy: userId},
+			}
+			err := impl.externalLinkClusterMappingRepository.Save(externalLinksMapping)
+			if err != nil {
+				impl.logger.Errorw("error in saving cluster id's", "data", t, "err", err)
+				err = &util.ApiError{
+					InternalMessage: "cluster id failed to create in db",
+					UserMessage:     "cluster id failed to create in db",
+				}
+				return nil, err
+			}
+		}
+	}
+	externalLinksCreateUpdateResponse := &ExternalLinkApiResponse{
+		Success: true,
+	}
+	return externalLinksCreateUpdateResponse, nil
+}
+
+func (impl ExternalLinkServiceImpl) GetAllActiveTools() ([]ExternalLinkMonitoringToolDto, error) {
+	impl.logger.Debug("fetch all links from db")
+	tools, err := impl.externalLinkMonitoringToolRepository.FindAllActive()
+	if err != nil {
+		impl.logger.Errorw("error in fetch all tools", "err", err)
+		return nil, err
+	}
+	var toolRequests []ExternalLinkMonitoringToolDto
+	for _, tool := range tools {
+		providerRes := ExternalLinkMonitoringToolDto{
+			Id:   tool.Id,
+			Name: tool.Name,
+			Icon: tool.Icon,
+		}
+		toolRequests = append(toolRequests, providerRes)
+	}
+	return toolRequests, err
+}
+
+func (impl ExternalLinkServiceImpl) FetchAllActiveLinks(clusterId int) ([]*ExternalLinkDto, error) {
+	impl.logger.Debug("fetch all links from db")
+	var err error
+	var mappedExternalLinksIds []int
+	filterByCluster := make(map[int]int)
+	externalLinksMap := make(map[int]int)
+	allActiveExternalLinks, err := impl.externalLinkClusterMappingRepository.FindAllActive()
+	for _, link := range allActiveExternalLinks {
+		externalLinksMap[link.ExternalLinkId] = link.ExternalLinkId
+	}
+	for _, externalLinksId := range externalLinksMap {
+		mappedExternalLinksIds = append(mappedExternalLinksIds, externalLinksId)
+	}
+
+	var externalLinkResponse []*ExternalLinkDto
+	response := make(map[int]*ExternalLinkDto)
+	for _, link := range allActiveExternalLinks {
+
+		//requested all links
+		if clusterId > 0 {
```

Comment:
logic can be more simpler

Reply:
this is simpler and fast, we can sit over alternative you have.